File: tmp/crop-lovart-grid.py
#!/usr/bin/env python3
import json
import math
import os
import subprocess
from PIL import Image

ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
SOURCE = os.path.join(ROOT, "apps/mini-program/assets-source/lovart/mini-program UI:status icons grid/mini-program UI:status icons grid.png")
BASE_OUT = os.path.join(ROOT, "apps/mini-program/src/assets/icons")

# 5x5 grid mapping (row, col) -> output tier/name
GRID = [
    [  # Row 1
        {"tier": "status", "name": "status-alarm-clock", "emoji": "⏰"},
        {"tier": "status", "name": "status-megaphone", "emoji": "📣"},
        {"tier": "status", "name": "status-bar-chart", "emoji": "📊"},
        {"tier": "status", "name": "status-warning", "emoji": "⚠️"},
        {"tier": "status", "name": "status-prohibited", "emoji": "🚫"},
    ],
    [  # Row 2
        {"tier": "semantic", "name": "label-cityscape", "emoji": "🌆"},
        {"tier": "semantic", "name": "label-map", "emoji": "🗺️"},
        {"tier": "semantic", "name": "label-globe-asia", "emoji": "🌏"},
        {"tier": "semantic", "name": "label-globe-meridians", "emoji": "🌐"},
        {"tier": "semantic", "name": "label-airplane", "emoji": "✈️"},
    ],
    [  # Row 3
        {"tier": "reaction", "name": "reaction-money-bag", "emoji": "💰"},
        {"tier": "reaction", "name": "reaction-smirk", "emoji": "😏"},
        {"tier": "reaction", "name": "reaction-sunglasses", "emoji": "😎"},
        {"tier": "reaction", "name": "reaction-purple-heart", "emoji": "💜"},
        {"tier": "reaction", "name": "reaction-sweat", "emoji": "😅"},
    ],
    [  # Row 4
        {"tier": "status", "name": "status-mirror", "emoji": "🪞"},
        {"tier": "status", "name": "status-unlocked", "emoji": "🔓"},
        {"tier": "status", "name": "status-star", "emoji": "🌟"},
        {"tier": "status", "name": "status-close", "emoji": "✕"},
        {"tier": "status", "name": "status-check", "emoji": "✓"},
    ],
    [  # Row 5
        {"tier": "reaction", "name": "reaction-devil", "emoji": "😈"},
        {"tier": "status", "name": "status-bell", "emoji": "🔔"},
        {"tier": "ui", "name": "icon-gift", "emoji": "🎁"},
        {"tier": "ui", "name": "icon-search", "emoji": "🔍"},
        {"tier": "ui", "name": "icon-memo", "emoji": "📝"},
    ],
]

TIER_FOLDER = {
    "status": "status-icons",
    "semantic": "info-labels",
    "reaction": "reaction-icons",
    "ui": "ui",
}

TARGET_SIZES = [96, 192, 288]
ALPHA_THRESHOLD = 20


def find_bounding_box_and_centroid(cell):
    data = cell.load()
    w, h = cell.size
    min_x, min_y = w, h
    max_x = max_y = -1
    sum_x = sum_y = total = 0
    for y in range(h):
        for x in range(w):
            alpha = data[x, y][3]
            if alpha > ALPHA_THRESHOLD:
                if x < min_x:
                    min_x = x
                if x > max_x:
                    max_x = x
                if y < min_y:
                    min_y = y
                if y > max_y:
                    max_y = y
                sum_x += x * alpha
                sum_y += y * alpha
                total += alpha
    if total == 0:
        return None
    bbox = (min_x, min_y, max_x + 1, max_y + 1)
    centroid = (sum_x / total, sum_y / total)
    return bbox, centroid


def main():
    img = Image.open(SOURCE).convert("RGBA")
    W, H = img.size
    rows = len(GRID)
    cols = len(GRID[0])

    x_boundaries = [round(i * W / cols) for i in range(cols + 1)]
    y_boundaries = [round(i * H / rows) for i in range(rows + 1)]

    report = []
    verification = img.copy()

    for r in range(rows):
        for c in range(cols):
            meta = GRID[r][c]
            left = x_boundaries[c]
            right = x_boundaries[c + 1]
            top = y_boundaries[r]
            bottom = y_boundaries[r + 1]
            cell_w = right - left
            cell_h = bottom - top

            cell = img.crop((left, top, right, bottom))
            result = find_bounding_box_and_centroid(cell)
            if result is None:
                report.append({**meta, "row": r, "col": c, "error": "empty cell"})
                print(f"[EMPTY] {meta['emoji']} @ row {r}, col {c}")
                continue

            (min_x, min_y, max_x, max_y), (cx, cy) = result
            bbox_w = max_x - min_x
            bbox_h = max_y - min_y
            bbox_cx = (min_x + max_x) / 2
            bbox_cy = (min_y + max_y) / 2

            # Square crop centered on centroid with 12% padding
            side = max(bbox_w, bbox_h) * 1.12
            side = min(side, cell_w, cell_h)
            crop_left = round(cx - side / 2)
            crop_top = round(cy - side / 2)
            if crop_left < 0:
                crop_left = 0
            if crop_top < 0:
                crop_top = 0
            if crop_left + side > cell_w:
                crop_left = round(cell_w - side)
            if crop_top + side > cell_h:
                crop_top = round(cell_h - side)
            side = round(side)

            cropped = cell.crop((crop_left, crop_top, crop_left + side, crop_top + side))

            folder = os.path.join(BASE_OUT, TIER_FOLDER[meta["tier"]])
            os.makedirs(folder, exist_ok=True)

            outputs = []
            for size in TARGET_SIZES:
                suffix = "" if size == 96 else f"@{size // 96}x"
                png_path = os.path.join(folder, f"{meta['name']}{suffix}.png")
                webp_path = os.path.join(folder, f"{meta['name']}{suffix}.webp")

                resized = cropped.resize((size, size), Image.Resampling.LANCZOS)
                resized.save(png_path, "PNG")

                quality = 90 if size == 96 else 85
                subprocess.run(
                    ["cwebp", "-q", str(quality), "-alpha_q", "100", png_path, "-o", webp_path],
                    check=True,
                    stdout=subprocess.DEVNULL,
                    stderr=subprocess.DEVNULL,
                )
                os.remove(png_path)
                outputs.append(os.path.basename(webp_path))

            # Draw verification overlay
            overlay = verification.load()
            v_left = left + crop_left
            v_top = top + crop_top
            for i in range(side):
                # red crosshair / box
                overlay[v_left + i, v_top] = (255, 0, 0, 255)
                overlay[v_left + i, v_top + side - 1] = (255, 0, 0, 255)
                overlay[v_left, v_top + i] = (255, 0, 0, 255)
                overlay[v_left + side - 1, v_top + i] = (255, 0, 0, 255)
            center_x = v_left + side // 2
            center_y = v_top + side // 2
            for i in range(-side // 6, side // 6 + 1):
                overlay[center_x + i, center_y] = (255, 0, 0, 255)
                overlay[center_x, center_y + i] = (255, 0, 0, 255)

            report.append(
                {
                    **meta,
                    "row": r,
                    "col": c,
                    "cell": f"{cell_w}x{cell_h}",
                    "bbox": f"{bbox_w}x{bbox_h}",
                    "centroidOffset": {
                        "x": round(cx - bbox_cx, 2),
                        "y": round(cy - bbox_cy, 2),
                    },
                    "crop": f"{side}x{side}",
                    "outputs": outputs,
                }
            )

            off_x = cx - bbox_cx
            off_y = cy - bbox_cy
            ok = abs(off_x) <= 4 and abs(off_y) <= 4
            print(
                f"{'✅' if ok else '⚠️'} {meta['emoji']} {meta['name']} | "
                f"centroid offset {off_x:+.2f},{off_y:+.2f} | crop {side}x{side}"
            )

    verif_path = os.path.join(ROOT, "tmp", "lovart-grid-crop-verification.png")
    verification.save(verif_path, "PNG")

    report_path = os.path.join(ROOT, "tmp", "lovart-grid-crop-report.json")
    with open(report_path, "w", encoding="utf-8") as f:
        json.dump(report, f, indent=2, ensure_ascii=False)

    print(f"\nCropped {len([r for r in report if not r.get('error')])}/{rows * cols} icons.")
    print(f"Verification overlay: {verif_path}")
    print(f"Report: {report_path}")


if __name__ == "__main__":
    main()

File: tmp/crop-lovart-grid.mjs
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { execSync } from 'child_process'
import { createCanvas, loadImage } from 'canvas'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)

const sourcePath = path.resolve(__dirname, '../apps/mini-program/assets-source/lovart/mini-program UI:status icons grid/mini-program UI:status icons grid.png')
const baseOut = path.resolve(__dirname, '../apps/mini-program/src/assets/icons')

// 5x5 grid mapping (row, col) -> output tier/name
const grid = [
  // Row 1
  [
    { tier: 'status', name: 'status-alarm-clock', emoji: '⏰' },
    { tier: 'status', name: 'status-megaphone', emoji: '📣' },
    { tier: 'status', name: 'status-bar-chart', emoji: '📊' },
    { tier: 'status', name: 'status-warning', emoji: '⚠️' },
    { tier: 'status', name: 'status-prohibited', emoji: '🚫' },
  ],
  // Row 2
  [
    { tier: 'semantic', name: 'label-cityscape', emoji: '🌆' },
    { tier: 'semantic', name: 'label-map', emoji: '🗺️' },
    { tier: 'semantic', name: 'label-globe-asia', emoji: '🌏' },
    { tier: 'semantic', name: 'label-globe-meridians', emoji: '🌐' },
    { tier: 'semantic', name: 'label-airplane', emoji: '✈️' },
  ],
  // Row 3
  [
    { tier: 'reaction', name: 'reaction-money-bag', emoji: '💰' },
    { tier: 'reaction', name: 'reaction-smirk', emoji: '😏' },
    { tier: 'reaction', name: 'reaction-sunglasses', emoji: '😎' },
    { tier: 'reaction', name: 'reaction-purple-heart', emoji: '💜' },
    { tier: 'reaction', name: 'reaction-sweat', emoji: '😅' },
  ],
  // Row 4
  [
    { tier: 'status', name: 'status-mirror', emoji: '🪞' },
    { tier: 'status', name: 'status-unlocked', emoji: '🔓' },
    { tier: 'status', name: 'status-star', emoji: '🌟' },
    { tier: 'status', name: 'status-close', emoji: '✕' },
    { tier: 'status', name: 'status-check', emoji: '✓' },
  ],
  // Row 5
  [
    { tier: 'reaction', name: 'reaction-devil', emoji: '😈' },
    { tier: 'status', name: 'status-bell', emoji: '🔔' },
    { tier: 'ui', name: 'icon-gift', emoji: '🎁' },
    { tier: 'ui', name: 'icon-search', emoji: '🔍' },
    { tier: 'ui', name: 'icon-memo', emoji: '📝' },
  ],
]

const tierFolder = {
  status: 'status-icons',
  semantic: 'info-labels',
  reaction: 'reaction-icons',
  ui: 'ui',
}

const targetSizes = [96, 192, 288]

async function main() {
  const img = await loadImage(sourcePath)
  const W = img.width
  const H = img.height
  const rows = grid.length
  const cols = grid[0].length

  // Build column / row boundaries (round to nearest pixel, full coverage)
  const xBoundaries = Array.from({ length: cols + 1 }, (_, i) => Math.round((i * W) / cols))
  const yBoundaries = Array.from({ length: rows + 1 }, (_, i) => Math.round((i * H) / rows))

  const report = []
  const verificationCanvas = createCanvas(W, H)
  const vCtx = verificationCanvas.getContext('2d')
  vCtx.drawImage(img, 0, 0)

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      const meta = grid[r][c]
      const left = xBoundaries[c]
      const right = xBoundaries[c + 1]
      const top = yBoundaries[r]
      const bottom = yBoundaries[r + 1]
      const cellW = right - left
      const cellH = bottom - top

      // Create a temporary canvas for the cell to read pixels
      const cellCanvas = createCanvas(cellW, cellH)
      const cellCtx = cellCanvas.getContext('2d')
      cellCtx.drawImage(img, left, top, cellW, cellH, 0, 0, cellW, cellH)
      const { data } = cellCtx.getImageData(0, 0, cellW, cellH)

      // Find bounding box of non-transparent pixels
      let minX = cellW, minY = cellH, maxX = -1, maxY = -1
      let alphaSumX = 0, alphaSumY = 0, alphaTotal = 0
      for (let y = 0; y < cellH; y++) {
        for (let x = 0; x < cellW; x++) {
          const idx = (y * cellW + x) * 4
          const alpha = data[idx + 3]
          if (alpha > 20) {
            if (x < minX) minX = x
            if (x > maxX) maxX = x
            if (y < minY) minY = y
            if (y > maxY) maxY = y
            alphaSumX += x * alpha
            alphaSumY += y * alpha
            alphaTotal += alpha
          }
        }
      }

      if (alphaTotal === 0) {
        report.push({ ...meta, row: r, col: c, error: 'empty cell' })
        continue
      }

      const bboxW = maxX - minX + 1
      const bboxH = maxY - minY + 1
      const cx = alphaSumX / alphaTotal
      const cy = alphaSumY / alphaTotal
      const bboxCx = (minX + maxX) / 2
      const bboxCy = (minY + maxY) / 2

      // Square crop centered on centroid, with 12% padding
      let side = Math.max(bboxW, bboxH) * 1.12
      side = Math.min(side, cellW, cellH)
      let cropLeft = Math.round(cx - side / 2)
      let cropTop = Math.round(cy - side / 2)
      // Clip to cell bounds
      if (cropLeft < 0) cropLeft = 0
      if (cropTop < 0) cropTop = 0
      if (cropLeft + side > cellW) cropLeft = Math.round(cellW - side)
      if (cropTop + side > cellH) cropTop = Math.round(cellH - side)
      side = Math.round(side)

      // Crop square from cell
      const cropCanvas = createCanvas(side, side)
      const cropCtx = cropCanvas.getContext('2d')
      cropCtx.drawImage(cellCanvas, cropLeft, cropTop, side, side, 0, 0, side, side)

      // Output to tier folder
      const folder = path.join(baseOut, tierFolder[meta.tier])
      fs.mkdirSync(folder, { recursive: true })

      for (const size of targetSizes) {
        const suffix = size === 96 ? '' : `@${size / 96}x`
        const outPng = path.join(folder, `${meta.name}${suffix}.png`)
        const outWebp = path.join(folder, `${meta.name}${suffix}.webp`)

        const resized = createCanvas(size, size)
        const rCtx = resized.getContext('2d')
        rCtx.imageSmoothingEnabled = true
        rCtx.imageSmoothingQuality = 'high'
        rCtx.drawImage(cropCanvas, 0, 0, size, size)

        const buf = resized.toBuffer('image/png')
        fs.writeFileSync(outPng, buf)

        // Convert PNG to WebP with cwebp
        const quality = size === 96 ? 90 : 85
        execSync(`cwebp -q ${quality} -alpha_q 100 "${outPng}" -o "${outWebp}"`)
        fs.unlinkSync(outPng)
      }

      // Verification overlay on full image
      const vLeft = left + cropLeft
      const vTop = top + cropTop
      vCtx.strokeStyle = 'red'
      vCtx.lineWidth = 2
      vCtx.strokeRect(vLeft, vTop, side, side)
      vCtx.beginPath()
      vCtx.moveTo(vLeft + side / 2, vTop)
      vCtx.lineTo(vLeft + side / 2, vTop + side)
      vCtx.moveTo(vLeft, vTop + side / 2)
      vCtx.lineTo(vLeft + side, vTop + side / 2)
      vCtx.stroke()

      report.push({
        ...meta,
        row: r,
        col: c,
        cell: `${cellW}x${cellH}`,
        bbox: `${bboxW}x${bboxH}`,
        centroidOffset: {
          x: Number((cx - bboxCx).toFixed(2)),
          y: Number((cy - bboxCy).toFixed(2)),
        },
        crop: `${side}x${side}`,
        outputs: targetSizes.map((s) => `${meta.name}${s === 96 ? '' : `@${s / 96}x`}.webp`),
      })
    }
  }

  // Save verification overlay
  const verifPath = path.resolve(__dirname, '../tmp/lovart-grid-crop-verification.png')
  fs.writeFileSync(verifPath, verificationCanvas.toBuffer('image/png'))

  // Save report JSON
  const reportPath = path.resolve(__dirname, '../tmp/lovart-grid-crop-report.json')
  fs.writeFileSync(reportPath, JSON.stringify(report, null, 2))

  // Console summary
  console.log(`Cropped ${report.filter((r) => !r.error).length}/${rows * cols} icons.`)
  console.log(`Verification overlay saved to: ${verifPath}`)
  console.log(`Report saved to: ${reportPath}`)
  for (const r of report) {
    if (r.error) {
      console.log(`[EMPTY] ${r.emoji} @ row ${r.row}, col ${r.col}`)
    } else {
      const off = r.centroidOffset
      const ok = Math.abs(off.x) <= 4 && Math.abs(off.y) <= 4
      console.log(`${ok ? '✅' : '⚠️'} ${r.emoji} ${r.name} | centroid offset ${off.x},${off.y} | crop ${r.crop}`)
    }
  }
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
